### PR TITLE
Add getNextHandlePosition for complex drag validation

### PR DESCRIPTION
**Pull request by tomlagier**
Addresses https://github.com/airbnb/rheostat/issues/125. Allows us to deny a move depending on the handle index and next value.

### CODE REVIEW

File: src/Slider.jsx [hunk 1]
```diff
@@ -568,6 +570,8 @@ class Rheostat extends React.Component {
         ? handlePos[idx - 1] + handleDimensions
         : SliderConstants.PERCENT_EMPTY, // 0% is the lowest value
     );
+
+    return validateMove ? validateMove(idx, nextPos) : nextPos;
```
Comment (ljharb) [inline src/Slider.jsx]: I don't think we should proxy through the return value from `validateMove`; we should be sanitizing and validating it somehow.

Comment (tomlagier) [reply src/Slider.jsx]: We can parseFloat the return to ensure that we're getting back a valid percent and then run it through the main body of validatePosition. Does that work?

If we get NaN back from parseFloat(validateMove), should we throw, discard it and use the original, or do something else?

Comment (ljharb) [reply src/Slider.jsx]: I'd expect an exception.

Comment (ljharb) [reply src/Slider.jsx]: However, I'd also expect a function named "validate" to return either true/false, or `null`/an array of errors/an error - not a float.

Comment (tomlagier) [reply src/Slider.jsx]: Both of the existing "validate" functions in the slider return corrected values, that's what I was basing my naming off of. Open to naming suggestions, and open to renaming the existing methods for clarity as well.

Comment (tomlagier) [reply src/Slider.jsx]: We could rename all `validate` to `adjust` -> `adjustValues`, `adjustPosition` and `adjustMove`?

**Comment by tomlagier** [comment]
Friendly ping here. What are your thoughts about merging, or renaming and merging?

**Comment by tomlagier** [comment]
Any updates here?

File: test/slider-test.jsx [hunk 2]
```diff
@@ -574,6 +574,19 @@ describe('Slider API', () => {
       assert(slider.validatePosition(0, 90) === 75, 'the handle reached its own max');
       assert(slider.validatePosition(1, 20) === 25, 'the handle reached its own min');
     });
+
+    it('should honor validateMove precondition', () => {
```
Comment (goatslacker) [inline test/slider-test.jsx]: Can we add a test case here that covers the error branch?

**Comment by ljharb** [comment]
Also, this will need a rebase so as to remove all merge commits.

File: test/slider-test.jsx (hunk 2 above)
Comment (tomlagier) [reply test/slider-test.jsx]: Test case added

**Comment by tomlagier** [comment]
Squashed the commits and added an error-path test case, let me know if I can do anything else.

**Comment by goatslacker** [comment]
@tomlagier travis is failing because:

```
/home/travis/build/airbnb/rheostat/test/slider-test.jsx
  602:66  error  Missing trailing comma  comma-dangle
  614:58  error  Missing trailing comma  comma-dangle
```

**Comment by tomlagier** [comment]
Whoops, hadn't seen the dangling comma on function arguments rule before. Should lint clean now.

File: src/Slider.jsx [hunk 3]
```diff
@@ -565,14 +566,35 @@ class Rheostat extends React.Component {
...
+    let nextPosition = proposedPosition;
+    if (validateMove) {
+      nextPosition = parseFloat(validateMove(idx, proposedPosition));
+
+      if (
+        Number.isNaN(nextPosition)
+        || nextPosition < SliderConstants.PERCENT_EMPTY
+        || nextPosition > SliderConstants.PERCENT_FULL
+      ) {
+        throw new Error('validateMove returned invalid value. Valid values are floats between 0 and 100');
```
Comment (ljharb) [inline src/Slider.jsx]: this should be a `TypeError`. also, let's change this text from "value" to "position".

**Comment by tomlagier** [comment]
Updated with requested changes.

**Pull request merged**